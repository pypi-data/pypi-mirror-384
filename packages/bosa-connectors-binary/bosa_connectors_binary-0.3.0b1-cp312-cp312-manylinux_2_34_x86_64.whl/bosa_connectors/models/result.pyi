from pydantic import BaseModel

class ActionResult(BaseModel):
    """Model for an action result."""
    success: bool
    message: str
