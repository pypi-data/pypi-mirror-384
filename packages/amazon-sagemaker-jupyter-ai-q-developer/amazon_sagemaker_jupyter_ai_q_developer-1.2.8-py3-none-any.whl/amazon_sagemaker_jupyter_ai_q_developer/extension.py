import os

from jupyter_server.extension.application import ExtensionApp
from tornado.web import StaticFileHandler

from amazon_sagemaker_jupyter_ai_q_developer.request_logger import (
    init_api_operation_logger,
)

from .handlers import TelemetryHandler
from .provider import AmazonQPersona

AMAZON_Q_AVATAR_PATH = str(os.path.join(os.path.dirname(__file__), "static", "q.svg"))


class SagemakerQChatExtension(ExtensionApp):
    name = "amazon_sagemaker_jupyter_ai_q_developer"
    handlers = [
        # serve the default persona avatar at this path.
        # the `()` at the end of the URL denotes an empty regex capture group,
        # required by Tornado.
        (
            rf"{AmazonQPersona.avatar_route}()",
            StaticFileHandler,
            {"path": AMAZON_Q_AVATAR_PATH},
        ),
        (r"amazon_sagemaker_jupyter_ai_q_developer/telemetry/?", TelemetryHandler),
    ]

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        init_api_operation_logger(self.log)
