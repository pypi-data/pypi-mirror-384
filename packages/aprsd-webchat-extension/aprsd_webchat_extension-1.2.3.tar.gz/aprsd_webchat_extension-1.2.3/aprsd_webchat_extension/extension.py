from aprsd_webchat_extension.cmds import webchat  # noqa: F401
