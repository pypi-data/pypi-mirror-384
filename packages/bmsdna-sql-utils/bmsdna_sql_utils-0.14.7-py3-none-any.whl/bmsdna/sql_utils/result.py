import dataclasses
from typing import TYPE_CHECKING, Any, List, Sequence, Tuple, Type, TypeVar, Union, overload

if TYPE_CHECKING:
    import pydantic

SomeDataClass = TypeVar("SomeDataClass")
Row = Tuple


@overload
def make_class_from_cursor(classtype: type[dict], description: Sequence[Tuple], values: Sequence) -> list[dict]: ...


@overload
def make_class_from_cursor(classtype: type[dict], description: Sequence[Tuple], values: tuple) -> dict: ...


@overload
def make_class_from_cursor(classtype: type, description: Sequence[Tuple], values: None) -> None: ...


@overload
def make_class_from_cursor(
    classtype: type[SomeDataClass], description: Sequence[Tuple], values: Sequence[Row]
) -> List[SomeDataClass]: ...


@overload
def make_class_from_cursor(
    classtype: type[SomeDataClass], description: Sequence[Tuple], values: Row
) -> SomeDataClass: ...


def make_class_from_cursor(
    classtype: type[SomeDataClass],
    description: Sequence[Tuple],
    values: Union[Sequence[Row], Row, None],
) -> Union[SomeDataClass, List[SomeDataClass], None]:
    if values is None:
        return None

    def get_index(name: str):
        index = -1
        for d in description:
            index += 1
            if d[0] == name:
                return index
        return -1

    def get_vl(name: str, type: Any):
        ind = get_index(name)
        if ind == -1:
            return None
        return values[ind]

    if type(values) == tuple or type(values).__name__ == "Row" or type(values).__name__ == "pyodbc.Row":
        if classtype == dict:
            names = [d[0] for d in description]
            return {name: get_vl(name, None) for name in names}  # type: ignore
        elif dataclasses.is_dataclass(classtype):
            prms = {f.name: get_vl(f.name, f.type) for f in dataclasses.fields(classtype)}  # type: ignore
            return classtype(**prms)
        elif hasattr(classtype, "model_validate"):  # Pydantic 2.0
            mod: "Type[pydantic.BaseModel]" = classtype  # type: ignore
            prms = {f: get_vl(f, None) for f in mod.model_fields.keys()}
            return mod.model_validate(prms)  # type: ignore
        else:
            import pydantic

            adapt = pydantic.TypeAdapter(classtype)
            names = [d[0] for d in description]
            dct = {name: get_vl(name, None) for name in names}
            return adapt.validate_python(dct)  # type: ignore
    else:
        return [make_class_from_cursor(classtype, description, item) for item in values]  # type: ignore
