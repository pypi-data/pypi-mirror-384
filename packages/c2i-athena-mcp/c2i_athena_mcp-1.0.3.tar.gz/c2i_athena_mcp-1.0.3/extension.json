{
  "name": "c2i-athena-mcp",
  "displayName": "C2i Athena MCP",
  "description": "AWS Athena database operations and MRD calculations for C2i Genomics",
  "version": "1.0.0",
  "author": "C2i Genomics",
  "homepage": "https://github.com/c2i-genomics/c2i-athena-mcp",
  "license": "MIT",
  "keywords": ["genomics", "athena", "aws", "mrd", "bioinformatics"],
  
  "main": {
    "command": "uvx",
    "args": ["c2i-athena-mcp"]
  },
  
  "configuration": {
    "title": "C2i Athena MCP Configuration",
    "type": "object",
    "properties": {
      "awsRegion": {
        "type": "string",
        "title": "AWS Region",
        "description": "AWS region for Athena operations",
        "default": "us-east-1",
        "enum": ["us-east-1", "us-west-2", "eu-west-1"]
      },
      "awsProfile": {
        "type": "string", 
        "title": "AWS Profile",
        "description": "AWS profile name for authentication (required)",
        "examples": ["C2iAlefResearchAccess-340686126008", "default"]
      },
      "s3OutputUri": {
        "type": "string",
        "title": "S3 Output URI",
        "description": "S3 URI for Athena query results",
        "default": "s3://airflow-research-main-1-us-east-1/athena/results/",
        "pattern": "^s3://[a-z0-9.-]+(/.*)?$"
      }
    },
    "required": ["awsProfile"]
  },
  
  "environmentMapping": {
    "AWS_REGION": "awsRegion",
    "AWS_PROFILE": "awsProfile", 
    "AWS_S3_OUTPUT_URI": "s3OutputUri"
  },
  
  "capabilities": {
    "tools": [
      {
        "name": "list_databases",
        "description": "List available Athena databases",
        "examples": ["Show me all databases", "List databases"]
      },
      {
        "name": "list_tables",
        "description": "List tables in a database",
        "examples": ["Show tables in research_db", "What tables are available?"]
      },
      {
        "name": "execute_query",
        "description": "Execute SQL queries against Athena",
        "examples": ["SELECT * FROM samples LIMIT 10", "Query the latest results"]
      },
      {
        "name": "calculate_mrd_score_for_sample", 
        "description": "Calculate MRD score for a sample",
        "examples": ["Calculate MRD for SAMPLE123", "What's the MRD score for this sample?"]
      },
      {
        "name": "calculate_mrd_for_dataset",
        "description": "Process MRD calculations for datasets",
        "examples": ["Calculate MRD for all samples in batch_001", "Process MRD for the dataset"]
      }
    ]
  },
  
  "installation": {
    "requirements": {
      "python": ">=3.11",
      "system": ["aws-cli"]
    },
    "preInstall": [
      "Ensure AWS CLI is configured with appropriate credentials",
      "Verify access to C2i Genomics AWS environment",
      "Check that the specified AWS profile exists"
    ],
    "postInstall": [
      "Test connection with: uvx c2i-athena-mcp --help",
      "Verify AWS permissions for Athena and S3 access"
    ]
  },
  
  "documentation": {
    "quickStart": "https://github.com/c2i-genomics/c2i-athena-mcp#quick-start",
    "examples": "https://github.com/c2i-genomics/c2i-athena-mcp#examples",
    "troubleshooting": "https://github.com/c2i-genomics/c2i-athena-mcp#troubleshooting"
  }
}