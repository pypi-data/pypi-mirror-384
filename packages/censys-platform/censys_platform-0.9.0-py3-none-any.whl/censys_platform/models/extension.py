"""Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT."""

from __future__ import annotations
from censys_platform.types import BaseModel
from typing import Optional
from typing_extensions import NotRequired, TypedDict


class ExtensionTypedDict(TypedDict):
    critical: NotRequired[bool]
    id: NotRequired[str]
    value: NotRequired[str]


class Extension(BaseModel):
    critical: Optional[bool] = None

    id: Optional[str] = None

    value: Optional[str] = None
