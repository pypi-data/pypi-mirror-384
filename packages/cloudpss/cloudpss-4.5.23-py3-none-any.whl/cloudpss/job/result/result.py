

import sys
from deprecated import deprecated
from ..messageStreamSender import MessageStreamSender
from ..messageStreamReceiver import MessageStreamReceiver

class Result(object):
    """
        结果视图基类，提供基础的结果获取接口
    """
    def __init__(self, receiver:MessageStreamReceiver,sender:MessageStreamSender=None) -> None:
        """
            初始化
        """
        self.result = {}
        self._receiver = receiver
        self._sender = sender
        self.__logsIndex = 0
        
    
    async def receive(self):
        async for msg in self._receiver:
            yield msg
            
    def __len__(self):
        return len(self._receiver)

    def __iter__(self):
        return self
    def __next__(self):
        return next(self._receiver)
        
    def __deepModify(self, dict1, dict2):

        for key, val in dict1.items():
            if type(val) is dict:
                if type(dict2) is dict and dict2.get(key, None) is None:
                    dict2[key] = val
                else:
                    self.__deepModify(val, dict2[key])
            else:
                dict2[key] = val
                
    def modify(self, data, model):
        """
            通过指定消息修改算例文件

            :params: data 消息字典 {}
            :params: model  项目

            >>> message= view.modify(data,model)

        """
        modifyData = data['data']
        payload = modifyData['payload']
        self.__deepModify(payload, model)
    
    def getMessagesByKey(self, key):
        """
            获取指定 key 的消息数据

            :params key: 数据key

            :returns: 对应 key 的数据数组

            >>> message= db.getMessagesByKey('log')
        """

        result = []
        for val in self._receiver.messages:
            if val.get('key', None) == key:
                result.append(val)
        return result

    def getMessagesByType(self, type):
        """
            获取指定类型的消息数据

            :params type: 数据类型

            :returns: 对应类型的数据数组

            >>> message= db.getMessagesByType('log')
        """
        result = []
        for val in self._receiver:
            if val['type'] == type:
                result.append(val)
        return result

    def getMessage(self, index):
        """
            获取指定位置的消息数据

            :params index: 数据的位置信息

            :returns: 消息数据

            >>> message= db.getMessage(1)
        """
        return self._receiver.messages[index]

    def getMessages(self):
        """
            获取所有消息数据

            :returns: 消息数据数组
        """
        return self._receiver.messages
    
    
    def getLogs(self):
        '''
            获取当前任务的日志

            >>>logs= result.getLogs()
            {...}
        '''
        result = []
        length = len(self._receiver.messages)
        if (length > self.__logsIndex):
            for num in range(self.__logsIndex, length):
                val = self.getMessage(num)
                if val['type'] == 'log':
                    result.append(val)
            self.__logsIndex = length
        return result
    
    def getMessageLength(self):
        """
            获取消息数据的长度

            :returns: 消息数据的长度
        """
        return len(self._receiver.messages)
        
    
    def waitFor(self,timeOut=sys.maxsize):
        """
            阻塞方法，直到任务完成

            :params timeOut: 超时时间
        """
        return self._receiver.waitFor(timeOut)
    
    @property
    @deprecated(version='3.0', reason="该方法将在 5.0 版本移除")
    def db(self):
        """
            获取数据库对象

            :returns: 数据库对象
        """
        return self._receiver
    
    def pop(self,index=-1):
        """
            pop 出缓存中的消息

            :returns: 消息数据
        """
        return self._receiver.messages.pop(index)