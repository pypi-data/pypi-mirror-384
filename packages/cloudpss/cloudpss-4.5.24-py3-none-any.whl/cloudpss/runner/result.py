import copy
import re
import collections

from cloudpss.utils.httprequests import request
from .storage import Storage

class Result(object):
    """
        结果处理类，从消息存储库中获取数据，并进行简单的整理

        可迭代器，迭代时按接收顺序返回数据

        >>> for data in result:
                print(data)

        也可以从类的 db 变量，获取数据存储类实例进行操作
    """
    def __init__(self, db):
        """
            初始化
        """
        self.n = 0
        self.db = db
        self.__logsIndex = 0

    @classmethod
    def load(cls, filePath):
        """
            加载本地结果文件

            :params: file 文件目录

            :return: 返回一个结果实例

            >>> result = Result.load('C:\\Users\\dps-dm\\cloudpss-sdk\\result\\424111.cjob')

        """
        db = Storage.load(filePath)

        return cls(db)

    @staticmethod
    def dump(result, file):
        '''
            保存结果到本地文件

            :params: file 保存文件的目录

            >>> Result.dump(file)
            {...}
        '''

        Storage.dump(result.db, file)

    def __iter__(self):
        return self

    def __next__(self):
        maxLength = self.db.getMessageLength()
        if self.n < maxLength:
            message = self.db.getMessage(self.n)
            self.n += 1
            return message
        raise StopIteration()

    def getLogs(self):
        '''
            获取当前任务的日志

            >>>logs= result.getLogs()
            {...}
        '''
        result = []
        length = self.db.getMessageLength()
        if (length > self.__logsIndex):
            for num in range(self.__logsIndex, length):
                val = self.db.getMessage(num)
                if val['type'] == 'log':
                    result.append(val)
            self.__logsIndex = length
        return result

    def __deepModify(self, dict1, dict2):

        for key, val in dict1.items():
            
                
            if type(val) is dict:
                if type(dict2) is dict and dict2.get(key, None) is None:
                    dict2[key] = val
                else:
                    self.__deepModify(val, dict2[key])
            else:
                dict2[key] = val

    def modify(self, data, model):
        """
            通过指定消息修改算例文件

            :params: data 消息字典 {}
            :params: model  项目

            >>> message= result.modify(data,model)

        """
        modifyData = data['data']
        payload = modifyData['payload']
        self.__deepModify(payload, model)

    def getMessagesByType(self, type):
        """
            获取指定类型的消息数据

            

            >>> message= result.getMessagesByType('log')
        """
        return self.db.getMessagesByType(type)


class EMTResult(Result):
    """
        电磁暂态结果处理类，继承 Result， 

        提供快捷 plot 数据的接口函数，获取到的 plot 数据为合并后的数据格式，不在是接收时分段的数据

        该类只提供 EMT 仿真使用

    """
    def __init__(self, *args, **kwargs):
        """
            初始化
        """
        Result.__init__(self, *args, **kwargs)
        self.result = {}
        self.__plotIndex = 0

    def getPlots(self):
        '''
            获取所有的 plots 数据

            >>> result.getPlots()
            {...}
        '''

        length = self.db.getMessageLength()
        if (length > self.__plotIndex):
            for num in range(self.__plotIndex, length):
                val = self.db.getMessage(num)
                if val['type'] == 'plot':
                    key = val['key']
                    if self.result.get(key, None) is None:
                        self.result[key] = copy.deepcopy(val)
                    else:
                        traces = val['data']['traces']
                        for i in range(len(traces)):
                            v = traces[i]
                            self.result[key]['data']['traces'][i]['x'].extend(
                                v['x'])
                            self.result[key]['data']['traces'][i]['y'].extend(
                                v['y'])
            self.__plotIndex = length
        return self.result.values()

    def getPlot(self, index: int):
        '''
            获取指定序号的数据分组

            :params: index 图表位置

            >>> result.getPlot(0)
            {...}
        '''
        self.getPlots()
        if self.result is not None:
            return self.result.get('plot-{0}'.format(int(index)), None)

    def getPlotChannelNames(self, index):
        '''
            获取一组输出分组下的所有通道名称

            :params: index 输出通道位置

            :return: 通道名称数组

            >>>names= result.getPlotChannelNames(0)
            []
        '''
        plot = self.getPlot(int(index))
        if plot is None:
            return None

        return [val['name'] for val in plot['data']['traces']]

    def getPlotChannelData(self, index, channelName):
        '''
            获取一组输出分组下指定通道名称的数据

            :params: index 输出通道位置
            :params: channelName 输出通道名称

            :return: 通道数据, 一个 trace 数据

            >>>channel= result.getPlotChannelData(0，'')
            {...}
        '''
        plot = self.getPlot(int(index))
        if plot is None:
            return None
        for val in plot['data']['traces']:
            if val['name'] == channelName:
                return val
        return val


class PowerFlowResult(Result):
    """
        潮流结果处理类，继承 Result， 

        提供快速获取 buses 和 branches 的接口，并提供潮流写入项目的接口

        该类只提供潮流仿真时使用

    """
    def __init__(self, *args, **kwargs):
        """
            初始化
        """
        Result.__init__(self, *args, **kwargs)

    def getBuses(self):
        """
            获取潮流结果 buses 数据

            >>> channel= result.getBuses()
            [...]
        """

        result = []
        pattern = re.compile("value=\"(.*)\"")
        data = self.db.getMessagesByKey('buses-table')
        if len(data) == 0:
            return []
        table = copy.deepcopy(data[0])
        columns = table['data']['columns']
        for column in columns:
            if column['type'] == 'html':
                data = column['data']
                for i in range(len(data)):
                    if data[i] != '':
                        s = pattern.findall(data[i])
                        data[i] = s[0].replace('/', '')

        result.append(table)
        return result

    def getBranches(self):
        """
            获取潮流结果 branches 数据

            >>> channel= result.getBranches()
            [...]
        """

        result = []
        pattern = re.compile("value=\"(.*)\"")
        data = self.db.getMessagesByKey('branches-table')
        if len(data) == 0:
            return []
        table = copy.deepcopy(data[0])

        columns = table['data']['columns']
        for column in columns:
            if column['type'] == 'html':
                data = column['data']
                for i in range(len(data)):
                    if data[i] != '':
                        s = pattern.findall(data[i])
                        data[i] = s[0].replace('/', '')

        result.append(table)
        return result

    def powerFlowModify(self, model):
        """
            潮流数据写入 model

            >>> channel= result.powerFlowModify(model)
        """

        data = self.db.getMessagesByKey('power-flow-modify')
        if len(data) == 0:
            raise Exception('未找到到数据')
        self.modify(data[0], model)


class IESResult(Result):
    """
        综合能源结果处理类，继承 Result， 

        提供快捷 plot 数据的接口函数，获取到的 plot 数据为合并后的数据格式，不在是接收时分段的数据

        该类只提供 IES 仿真使用

    """
    def __init__(self, *args, **kwargs):
        """
            初始化
        """
        Result.__init__(self, *args, **kwargs)
        self.result = {'Sankey': []}
        self.__plotIndex = 0

    def __readPlotResult(self):
        length = self.db.getMessageLength()
        if (length > self.__plotIndex):
            for num in range(self.__plotIndex, length):
                val = self.db.getMessage(num)
                if val['type'] == 'plot':
                    key = val['key']
                    if key == 'Sankey':
                        self.result['Sankey'].append(copy.deepcopy(val))
                    else:
                        if self.result.get(key, None) is None:
                            self.result[key] = copy.deepcopy(val)
                        else:
                            traces = val['data']['traces']
                            for i in range(len(traces)):
                                v = traces[i]
                                self.result[key]['data']['traces'][i][
                                    'x'].extend(v['x'])
                                self.result[key]['data']['traces'][i][
                                    'y'].extend(v['y'])
            self.__plotIndex = length

    def getPlotData(self, compID, labelName, traceName='all', index=-1):
        '''
            获取元件ID为compID的元件，对应标签为labelName、图例名称为traceName的plot 数据的第index项

            :params: compID string类型，代表元件的标识符
            :params: labelName string类型，代表plot曲线的分组标签
            :params: traceName string类型，代表Plot曲线对应分组下的图例名称，当为'all'时，返回所有图例的数据
            :params: index int类型，代表对应图例时序数据中的第index项，当小于0时，返回该图例所有的时序数据
            
            :return: dict类型
        '''
        self.__readPlotResult()
        key = compID + '_' + labelName
        if key not in self.result.keys():
            raise Exception('未找到元件标志为{0},对应label为{1}的plot数据'.format(
                compID, labelName))

        traceData = self.result[key]['data']['traces']
        if traceName != 'all':
            traceNameList = [traceName]
        else:
            traceNameList = [
                traceData[i]['name'] for i in range(len(traceData))
            ]

        startIndex = 0
        endIndex = len(traceData[0]['x'])
        if index >= 0:
            startIndex = index
            endIndex = index + 1

        plotData = collections.defaultdict(lambda: {})
        for tName in traceNameList:
            for i in range(len(traceData)):
                dataLen = len(traceData[i]['x'])
                if traceData[i]['name'] == tName:
                    if endIndex > dataLen:
                        raise Exception('请求的index超过了plot数据序列的长度')
                    plotData[tName]['x'] = traceData[i]['x'][
                        startIndex:endIndex]
                    plotData[tName]['y'] = traceData[i]['y'][
                        startIndex:endIndex]

        return plotData

    def getSankey(self, index):
        '''
            获取第index个桑基图数据

            >>> result.getSankey(index)
            {...}
        '''
        self.__readPlotResult()
        if index >= len(self.result['Sankey']):
            raise Exception('index超过了桑基图数据序列的长度')
        return self.result['Sankey'][index]

    def getSankeyNum(self):
        '''
            获取桑基图数据序列的长度

            >>> result.getSankeyNum()
        '''
        self.__readPlotResult()
        return len(self.result['Sankey'])

class IESLabSimulationResult(IESResult):
    pass

class IESLabTypicalDayResult(IESResult):
    def GetTypicalDayNum():
        '''
            获取当前result的典型日数量
            
            :return: int类型，代表典型日数量
        '''
    def GetTypicalDayInfo(dayID):
        '''
            获取dayID对应典型日的基础信息
            
            :params: dayID int类型，表示典型日的ID，数值位于 0~典型日数量 之间
            
            :return: dict类型，代表典型日的基础信息，包括典型日所代表的日期范围、典型日的名称等
        '''
    def GetTypicalDayCurve(dayID, dataType):
        '''
            获取dayID对应典型日下dataType参数的时序曲线
            
            :params: dayID int类型，表示典型日的ID，数值位于 0~典型日数量 之间
            :params: dataType enum类型，标识辐照强度、环境温度、土壤温度、建筑物高度风速、风机高度风速、电负荷、热负荷、冷负荷的参数类型
            
            :return: list<float>类型，代表以1h为时间间隔的该参数的日内时序曲线
        '''
    pass
