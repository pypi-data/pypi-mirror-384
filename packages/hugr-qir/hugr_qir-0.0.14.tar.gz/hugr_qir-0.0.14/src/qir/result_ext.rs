use anyhow::{Result, anyhow, bail};
use hugr::{
    HugrView, Node,
    extension::{prelude::ConstString, simple_op::MakeExtensionOp as _},
    ops::ExtensionOp,
};
use hugr_llvm::{
    emit::{EmitFuncContext, EmitOpArgs, emit_value},
    inkwell::types::BasicType as _,
    sum::LLVMSumValue,
    types::HugrSumType,
};
use tket_qsystem::extension::result::{ResultOp, ResultOpDef};

use super::QirCodegenExtension;
impl QirCodegenExtension {
    pub fn emit_result_op<'c, H: HugrView<Node = Node>>(
        &self,
        context: &mut EmitFuncContext<'c, '_, H>,
        args: EmitOpArgs<'c, '_, ExtensionOp, H>,
        op: ResultOpDef,
    ) -> Result<()> {
        let result_op = ResultOp::from_extension_op(&args.node())?;
        let tag_str = result_op.tag;
        if tag_str.is_empty() {
            bail!("Empty result tag received")
        }

        let i8_ptr_ty = context
            .iw_context()
            .i8_type()
            .ptr_type(Default::default())
            .as_basic_type_enum();
        let tag_ptr = {
            let x = emit_value(context, &ConstString::new(tag_str).into())?;
            if x.get_type() == i8_ptr_ty {
                x
            } else {
                context.builder().build_bit_cast(x, i8_ptr_ty, "tag_ptr")?
            }
        };

        match op {
            ResultOpDef::Bool => {
                let [val] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("result_bool expects one input"))?;
                let bool_type = context.llvm_sum_type(HugrSumType::new_unary(2))?;
                let val = LLVMSumValue::try_new(val, bool_type)
                    .map_err(|_| anyhow!("bool_type expects a value"))?
                    .build_get_tag(context.builder())?;
                let i1_ty = context.iw_context().bool_type();
                let trunc_val = context.builder().build_int_truncate(val, i1_ty, "")?;
                let print_fn_ty = context
                    .iw_context()
                    .void_type()
                    .fn_type(&[i1_ty.into(), i8_ptr_ty.into()], false);
                let print_fn =
                    context.get_extern_func("__quantum__rt__bool_record_output", print_fn_ty)?;
                context.builder().build_call(
                    print_fn,
                    &[trunc_val.into(), tag_ptr.into()],
                    "print_bool",
                )?;
                args.outputs.finish(context.builder(), [])
            }
            ResultOpDef::Int | ResultOpDef::UInt => {
                let [mut val] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("result_bool expects one input"))?;
                let i64_ty = context.iw_context().i64_type();
                if val.get_type() != i64_ty.into() {
                    val = if op == ResultOpDef::Int {
                        context
                            .builder()
                            .build_int_s_extend(val.into_int_value(), i64_ty, "")
                    } else {
                        context
                            .builder()
                            .build_int_z_extend(val.into_int_value(), i64_ty, "")
                    }?
                    .into();
                }
                let print_fn_ty = context
                    .iw_context()
                    .void_type()
                    .fn_type(&[i64_ty.into(), i8_ptr_ty.into()], false);
                let print_fn =
                    context.get_extern_func("__quantum__rt__int_record_output", print_fn_ty)?;
                context.builder().build_call(
                    print_fn,
                    &[val.into(), tag_ptr.into()],
                    "print_bool",
                )?;
                args.outputs.finish(context.builder(), [])
            }
            ResultOpDef::F64 => {
                let [val] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("result_bool expects one input"))?;
                let f64_ty = context.iw_context().f64_type();
                let print_fn_ty = context
                    .iw_context()
                    .void_type()
                    .fn_type(&[f64_ty.into(), i8_ptr_ty.into()], false);
                let print_fn =
                    context.get_extern_func("__quantum__rt__double_record_output", print_fn_ty)?;
                context.builder().build_call(
                    print_fn,
                    &[val.into(), tag_ptr.into()],
                    "print_bool",
                )?;
                args.outputs.finish(context.builder(), [])
            }
            // these ops are not supported yet
            ResultOpDef::ArrBool
            | ResultOpDef::ArrInt
            | ResultOpDef::ArrUInt
            | ResultOpDef::ArrF64
            | _ => bail!("Unknown op: {op:?}"),
        }
    }
}

#[cfg(test)]
mod test {
    use hugr::ops::OpType;
    use hugr_llvm::{
        check_emission,
        test::{TestContext, llvm_ctx},
    };
    use rstest::rstest;

    use tket_qsystem::extension::result::ResultOpDef;

    use crate::qir::{QirCodegenExtension, QirPreludeCodegen};
    use crate::test::single_op_hugr;

    #[rstest::fixture]
    fn ctx(mut llvm_ctx: TestContext) -> TestContext {
        llvm_ctx.add_extensions(|builder| {
            builder
                .add_extension(QirCodegenExtension)
                .add_prelude_extensions(QirPreludeCodegen)
                .add_default_int_extensions()
                .add_float_extensions()
        });
        llvm_ctx
    }

    #[rstest]
    #[case(ResultOpDef::F64.instantiate(&["foo_f64".into()]).unwrap())]
    #[case(ResultOpDef::UInt.instantiate(&["foo_uint".into(), 3.into()]).unwrap())]
    #[case(ResultOpDef::Int.instantiate(&["foo_int".into(), 4.into()]).unwrap())]
    #[case(ResultOpDef::Bool.instantiate(&["bool_int".into()]).unwrap())]
    fn emit(ctx: TestContext, #[case] op: impl Into<OpType>) {
        let op = op.into();
        let mut insta = insta::Settings::clone_current();
        insta.set_snapshot_suffix(format!("{}_{}", insta.snapshot_suffix().unwrap_or(""), op));
        insta.bind(|| {
            let hugr = single_op_hugr(op);
            check_emission!(hugr, ctx);
        })
    }
}
