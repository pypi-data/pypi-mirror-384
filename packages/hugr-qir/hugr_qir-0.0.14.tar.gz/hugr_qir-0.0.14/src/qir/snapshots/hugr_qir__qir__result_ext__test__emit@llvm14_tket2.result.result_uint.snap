---
source: src/qir/result_ext.rs
expression: mod_str
---
; ModuleID = 'test_context'
source_filename = "test_context"

@0 = private unnamed_addr constant [9 x i8] c"foo_uint\00", align 1

define void @_hl.main.1(i8 %0) {
alloca_block:
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  %1 = zext i8 %0 to i64
  call void @__quantum__rt__int_record_output(i64 %1, i8* getelementptr inbounds ([9 x i8], [9 x i8]* @0, i32 0, i32 0))
  ret void
}

declare void @__quantum__rt__int_record_output(i64, i8*)
